Refine README.md as self-contained Getting Started (#34)
openvinotoolkit/openvino_xai: README.md

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
[![PyPI](https://img.shields.io/pypi/v/openvino_xai)](https://pypi.org/project/openvino_xai)
[![Downloads](https://static.pepy.tech/badge/openvino_xai)](https://pepy.tech/project/openvino_xai)

---

</div>

![OpenVINO XAI Concept](docs/source/_static/ovxai-concept.svg)

**OpenVINO™ Explainable AI (XAI) Toolkit** provides a suite of XAI algorithms for visual explanation of
[**OpenVINO™**](https://github.com/openvinotoolkit/openvino) Intermediate Representation (IR) models.

Given **OpenVINO** models and input images, **OpenVINO XAI** generates **saliency maps**
which highlights regions of the interest in the inputs from the models' perspective
to help users understand the reason why the complex AI models output such responses.

---

## Features

### What's new in v1.0.0

* Support generation of classification and detection per-class and per-image saliency maps
* Enable White-Box ([ReciproCAM](https://arxiv.org/abs/2209.14074)) and Black-Box ([RISE](https://arxiv.org/abs/1806.07421v3)) eXplainable AI algorithms

> * Support CNNs and Transformer-based architectures (validation on diverse set of timm models)

Reviewer: ```suggestion
* Support CNNs and Transformer-based architectures (validation on diverse set of [timm](https://github.com/huggingface/pytorch-image-models) models)
```

> * Enable Explainer (stateful object) as the main interface for XAI algorithms

Reviewer: ```suggestion
* Enable `Explainer` (stateful object) as the main interface for XAI algorithms
```

| Domain          | Task                 | Type      | Algorithm           | Links |
|-----------------|----------------------|-----------|---------------------|-------|
| Comptuer Vision | Image Classification | Black-Box | RISE                | [arxiv](https://arxiv.org/abs/1806.07421v3) / [src](openvino_xai/methods/black_box/rise.py) |
|                 |                      | White-Box | ReciproCAM          | [arxiv](https://arxiv.org/abs/2209.14074) / [src](openvino_xai/methods/white_box/recipro_cam.py) |
|                 |                      |           | VITReciproCAM       | [arxiv](https://arxiv.org/abs/2310.02588) / [src](openvino_xai/methods/white_box/recipro_cam.py) |
|                 |                      |           | ActivationMap       | experimental / [src](openvino_xai/methods/white_box/activation_map.py) |
|                 | Object Detection     |           | ClassProbabilityMap | experimental / [src](openvino_xai/methods/white_box/det_class_probability_map.py) |

### Supported explainable models

Most of CNNs and Transformer models from [Pytorch Image Models (timm)](https://github.com/huggingface/pytorch-image-models) are supported and validated.

> Please refer to the following kwnon issues for unsupported models.

Reviewer: ```suggestion
Please refer to the following known issues for unsupported models.
```

which highlights regions of the interest in the inputs from the models' perspective
to help users understand the reason why the complex AI models output such responses.

---

## Features

### What's new in v1.0.0

* Support generation of classification and detection per-class and per-image saliency maps
* Enable White-Box ([ReciproCAM](https://arxiv.org/abs/2209.14074)) and Black-Box ([RISE](https://arxiv.org/abs/1806.07421v3)) eXplainable AI algorithms
* Support CNNs and Transformer-based architectures (validation on diverse set of [timm](https://github.com/huggingface/pytorch-image-models) models)
* Enable `Explainer` (stateful object) as the main interface for XAI algorithms
* Support `AUTO` mode by default to detect the best XAI method for given models
* Expose `insert_xai` functional API to support XAI head insertion for OpenVINO IR models

Please refer to the [change logs](CHANGELOG.md) for the full release history.

### Supported XAI methods

At the moment, *Image Classification* and *Object Detection* tasks are supported for the *Computer Vision* domain.
*Black-Box* (model agnostic but slow) methods and *White-Box* (model specific but fast) methods are supported:

> | Domain          | Task                 | Type      | Algorithm           | Links |
> |-----------------|----------------------|-----------|---------------------|-------|
> | Comptuer Vision | Image Classification | Black-Box | RISE                | [arxiv](https://arxiv.org/abs/1806.07421v3) / [src](openvino_xai/methods/black_box/rise.py) |
> |                 |                      | White-Box | ReciproCAM          | [arxiv](https://arxiv.org/abs/2209.14074) / [src](openvino_xai/methods/white_box/recipro_cam.py) |
> |                 |                      |           | VITReciproCAM       | [arxiv](https://arxiv.org/abs/2310.02588) / [src](openvino_xai/methods/white_box/recipro_cam.py) |
> |                 |                      |           | ActivationMap       | experimental / [src](openvino_xai/methods/white_box/activation_map.py) |
> |                 | Object Detection     |           | ClassProbabilityMap | experimental / [src](openvino_xai/methods/white_box/det_class_probability_map.py) |

Reviewer: ```suggestion
| Domain          | Task                 | Type      | Algorithm           | Links |
|-----------------|----------------------|-----------|---------------------|-------|
| Computer Vision | Image Classification | White-Box | ReciproCAM          | [arxiv](https://arxiv.org/abs/2209.14074) / [src](openvino_xai/methods/white_box/recipro_cam.py) |
|                 |                      |           | VITReciproCAM       | [arxiv](https://arxiv.org/abs/2310.02588) / [src](openvino_xai/methods/white_box/recipro_cam.py) |
|                 |                      |           | ActivationMap       | experimental / [src](openvino_xai/methods/white_box/activation_map.py) |
|                   |                    | Black-Box | RISE                | [arxiv](https://arxiv.org/abs/1806.07421v3) / [src](openvino_xai/methods/black_box/rise.py) |
|                 | Object Detection     |           | ClassProbabilityMap | experimental / [src](openvino_xai/methods/white_box/det_class_probability_map.py) |
```

Fix typo and white-box / black-box methods order

### Supported explainable models

Most of CNNs and Transformer models from [Pytorch Image Models (timm)](https://github.com/huggingface/pytorch-image-models) are supported and validated.

> Please refer to the following known issues for unsupported models.

Reviewer 2: This might be the main issue, let's put it on the first place
https://github.com/openvinotoolkit/openvino_xai/issues/26

Author: https://github.com/openvinotoolkit/openvino_xai/pull/34/commits/dd5776f6a040fd36580f81175d4702cae4a78943

# Activate virtual env.
source .ovxai/bin/activate
```
</details>

Install from PyPI package

```bash
# Base package (for normal use):
pip install openvino_xai

# Dev package (for development):
pip install openvino_xai[dev]
```

<details>
<summary>Install from source</summary>

```bash
# Clone the source repository
git clone https://github.com/openvinotoolkit/openvino_xai.git
cd openvino_xai

# Editable mode (for development):
pip install -e .[dev]
```
</details>

<details>
<summary>Verify installation</summary>

```bash
# Run tests
pytest -v -s ./tests/unit

# Run code quality checks
pre-commit run --all-files
```
</details>

---

## Quick Start

### Hello, OpenVINO XAI

Let's imagine the case that our OpenVINO IR model is up and running on a inference pipeline.
While watching the outputs, we may want to analyze the model's behavior for debugging or understanding purposes.

> By using the **OpenVINO XAI** `Explainer`, we can visualize why the model outputs such responses.

Reviewer: ```suggestion
By using the **OpenVINO XAI** `Explainer`, we can visualize why the model gives such responses.
```

> In this examples, we are trying to know the reason why the model outputs a "cheetah" label for given input image.

Reviewer: ```suggestion
In this examples, we are trying to know the reason why the model outputs a `cheetah` label for the given input image.
```

```python
import cv2
import numpy as np
import openvino.runtime as ov
import openvino_xai as xai

# Load the model
ov_model: ov.Model = ov.Core().read_model("mobilenet_v3.xml")

# Load the image to be analized
image: np.ndarray = cv2.imread("tests/assets/cheetah_person.jpg")
image = cv2.resize(image, dsize=(224, 224))
image = np.expand_dims(image, 0)

# Create the Explainer object
explainer = xai.Explainer(
    model=ov_model,
    task=xai.Task.CLASSIFICATION,
)

# Generate saliency map for the label of interest
explanation: xai.Explanation = explainer(
    data=image,

>     targets=293,  # (cheetah), accepts single or list of targets

Reviewer: ```suggestion
    targets=293,  # (cheetah), accepts single int or list of targets (ints or strings)
```

    overlay=True,  # saliency map overlay over the input image, defaults to False
)

# Save saliency maps to output directory
explanation.save(dir_path="./output")
```

Original image | Explained image
---------------|----------------
![Oringinal images](tests/assets/cheetah_person.jpg) | ![Explained image](docs/source/_static/xai-cheetah.png)

We can see that model is focusing on the body or skin area of the animals to tell if this image contains actual cheetahs.

### More advanced use-cases

Users could tweak the basic use-case according to their purpose, which include but not limited to:

* Select XAI mode (White-Box or Black-Box) or even specific method which are automatically decided by default
* Provide custom model pre/post processing functions like resize and normalizations which the model expects

> * Customize output image visualization options
> * Explain multiple class targets and images

Reviewer: ```suggestion
* Customize output image visualization options
* Explain multiple class targets, passing them as label indexes or as actual label names
* Explain multiple images
```

<summary>Verify installation</summary>

```bash
# Run tests
pytest -v -s ./tests/unit

# Run code quality checks
pre-commit run --all-files
```
</details>

---

## Quick Start

### Hello, OpenVINO XAI

Let's imagine the case that our OpenVINO IR model is up and running on a inference pipeline.
While watching the outputs, we may want to analyze the model's behavior for debugging or understanding purposes.

By using the **OpenVINO XAI** `Explainer`, we can visualize why the model gives such responses.
In this example, we are trying to know the reason why the model outputs a `cheetah` label for the given input image.

```python
import cv2
import numpy as np
import openvino.runtime as ov
import openvino_xai as xai

# Load the model
ov_model: ov.Model = ov.Core().read_model("mobilenet_v3.xml")

# Load the image to be analyzed
image: np.ndarray = cv2.imread("tests/assets/cheetah_person.jpg")
image = cv2.resize(image, dsize=(224, 224))
image = np.expand_dims(image, 0)

# Create the Explainer object
explainer = xai.Explainer(
    model=ov_model,
    task=xai.Task.CLASSIFICATION,
)

# Generate saliency map for the label of interest
explanation: xai.Explanation = explainer(
    data=image,

>     targets=293,  # (cheetah), accepts label indices or actual label names

Reviewer 2: ```suggestion
    targets=293,  # (cheetah), accepts label indices or actual label names if label_names provided
```

    overlay=True,  # saliency map overlay over the input image, defaults to False
)

# Save saliency maps to output directory
explanation.save(dir_path="./output")
```

Original image | Explained image
---------------|----------------
![Oringinal images](tests/assets/cheetah_person.jpg) | ![Explained image](docs/source/_static/xai-cheetah.png)

We can see that model is focusing on the body or skin area of the animals to tell if this image contains actual cheetahs.

### More advanced use-cases

Users could tweak the basic use-case according to their purpose, which include but not limited to:

* Select XAI mode (White-Box or Black-Box) or even specific method which are automatically decided by default
* Provide custom model pre/post processing functions like resize and normalizations which the model expects
* Customize output image visualization options
* Explain multiple class targets, passing them as label indices or as actual label names

> * Explain multiple images

Reviewer 2: We do not natively support multiple images, the user should call explainer multiple times. I would reformulate this:
```suggestion
* Call explainer multiple times to explain multiple images or to use different targets
```

> 

Reviewer 2: I would add `insert_xai` use case here:

Avoid OpenVINO XAI dependency in the inference environment by inserting XAI head into the model beforehand using `insert_xai`

Author: https://github.com/openvinotoolkit/openvino_xai/pull/34/commits/dd5776f6a040fd36580f81175d4702cae4a78943
